packages/core/test/core/plan/plan.test.ts

```diff
@@ -219,13 +230,212 @@ describe('getPlan', () => {
+      {
+        ref: { inner: 'some value' },
+      }
+    )
+
+    const secondInstance4 = new InstanceElement(
+      'instance4',
+      type,
+      {
+        ref: new ReferenceExpression(
+          new ElemID('adapter', 'type', 'instance', 'instance4', 'inner'),
+          {
+            inner: new ReferenceExpression(
+              new ElemID('adapter', 'type', 'instance', 'instance4', 'value'),
+              'some value'
+            ),
+          }
+        ),
+      }
+    )
+
+
+    const plan = await getPlan({
+      before: createElementSource(
+        [firstInstance1, firstInstance2, firstInstance3, firstInstance4, type, innerType]
+      ),
+      after: createElementSource(
+        [secondInstance1, secondInstance2, secondInstance3, secondInstance4, type, innerType]
+      ),
+    })
+    expect(plan.size).toBe(0)
+  })
+
+  it('when instances use variables and there is no change should create empty plan', async () => {
+    const variableObject = new Variable(
+      new ElemID('var', 'a'),
+      5
+    )
+
+    const type = new ObjectType({
+      elemID: new ElemID('adapter', 'type'),
+      fields: {
+        value: { refType: BuiltinTypes.NUMBER },
+      },
+    })
+
+    const instance = new InstanceElement(
+      'instance',
+      type,
+      {
+        value: new VariableExpression(variableObject.elemID),
+      }
+    )
+
+    const stateInstance = new InstanceElement(
+      'instance',
+      type,
+      {
+        value: 5,
+      }
+    )
+
     const plan = await getPlan({
-      before: createElementSource([stateInstance, type, innerType]),
-      after: createElementSource([instance, type, innerType]),
+      before: createElementSource([stateInstance, type]),
+      after: createElementSource([instance, type, variableObject]),
     })
     expect(plan.size).toBe(0)
   })
 
+  it('when reference in instance changes but the value is the same should have a change in plan', async () => {
+    const type = new ObjectType({
+      elemID: new ElemID('adapter', 'type'),
+      fields: {
+        a: { refType: BuiltinTypes.NUMBER },
+        b: { refType: BuiltinTypes.NUMBER },
+        ref: { refType: BuiltinTypes.NUMBER },
+      },
+    })
+
+    const instance = new InstanceElement(
+      'instance',
+      type,
+      {
+        a: 5,
+        b: 5,
+        ref: new ReferenceExpression(new ElemID('adapter', 'type', 'instance', 'instance', 'a')),
+      }
+    )
+
+    const changedInstance = new InstanceElement(
+      'instance',
+      type,
+      {
+        a: 5,
+        b: 5,
+        ref: new ReferenceExpression(new ElemID('adapter', 'type', 'instance', 'instance', 'b')),
+      }
+    )
+
+    const plan = await getPlan({
+      before: createElementSource([instance, type]),
+      after: createElementSource([changedInstance, type]),
+    })
+    expect(plan.size).toBe(1)
+  })
+
+  it('when reference in instance points to a whole element should have a change in plan', async () => {
```

Comment:
again, not really good behavior, but this one can't be solved with the current design unfortunately.
again, probably not your change so no need to change, but maybe a comment above the test?

```suggestion
  it('when reference in instance points to a whole element should have a change in plan', async () => {
    // Ideally we would want to know if the adapter is going to resolve this element into its ID so we could tell if this
    // is a real difference, but since we can't do that with the current design, we take the safer option and say this is
    // always a change
```